plugins/package-managers/node/src/main/kotlin/Npm.kt

```diff
@@ -246,8 +240,13 @@ open class Npm(
 
         logger.info { "Searching for 'package.json' files in '$nodeModulesDir'..." }
 
-        val nodeModulesFiles = nodeModulesDir.walk().maxDepth(modulesSearchDepth).filter {
-            it.name == "package.json" && isValidNodeModulesDirectory(nodeModulesDir, nodeModulesDirForPackageJson(it))
+        val visitedDirs = mutableSetOf<File>()
+
+        val nodeModulesFiles = nodeModulesDir.walk().onEnter { dir ->
+            !dir.isSymbolicLink() || visitedDirs.add(dir.realFile())
```

Comment:
I'm not sure if `!dir.isSymbolicLink()` is correct. Would it work to only work with "realDirs" like with the following (to be beautified) code:

```
onEnter { dir ->
    val realDir  = dir.realFile()
    if (realDir in visitedDirs) {
       false
    } else {
        visitedDirs += realDir
        true
    }
```

Reply:
> I'm not sure if `!dir.isSymbolicLink()` is correct.

What makes you doubt? IMO that's exactly the check we want to perform, and it's short and precise. Also, I've verified it to work with the project mentioned in #6531.

Reply:
In order to understand whether not following symlinks is correct, one needs to have a full understanding 
of 1. the link symlink structure of the multiple NPM package managers and 2. the directory structure.

Not proccessing a realDir twice in my view is (more) obvioulsy correct. (isn't that short an precise as well) ?

Reply:
> Not proccessing a realDir twice in my view is (more) obvioulsy correct.

My thinking is: You can never process a realdir twice if never following symbolic links. That is, the realdir check only needs to be performed when a symlink has been followed.

> isn't that short an precise as well

It's easy to phrase in prose, but you yourself admitted that [the code is not](https://github.com/oss-review-toolkit/ort/pull/8156#discussion_r1464948455).

Reply:
> My thinking is: You can never process a realdir twice if never following symbolic links. That is, the realdir check only needs to be performed when a symlink has been followed.

That's clear. Do you also have an explanation why you wouldn't miss a realDir, if not following symlinks?

Reply:
> Do you also have an explanation why you wouldn't miss a realDir, if not following symlinks?

I'm not sure I follow why the question is relevant. My code does not *never* follow symlinks, but it only follows a symlink if it does not point to a realdir that has been visited before (`Set.add()` returns `true` only if the realdir was actually added, i.e. not yet present in the set).

Reply:
Argh, sorry. I mis-read that expression in line 246. Now it's clear. Thanks!